### ISSUE TITLE
Qualifiers Ordering

### ISSUE THREAD

**Issue by tommyknows** [package-url/packageurl-go#53]
Hi again!

After #52, there seems to be a "bug" about the ordering of the qualifiers.
[The code says](https://github.com/package-url/packageurl-go/blob/master/packageurl.go#L120):
```go
// Qualifiers is a slice of key=value pairs, with order preserved as it appears
// in the package URL.
```

However, with the switch to the `url.URL` type, when encoding the qualifiers they actually [get ordered lexicographically](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/net/url/url.go;l=977).

Now I've started to wonder what the "correct" behaviour for this is. The [PURL Spec even says](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#how-to-build-purl-string-from-its-components):
> If the qualifiers are not empty and not composed only of key/value pairs where the value is empty:
> ...
> - sort this list of qualifier strings lexicographically
> ...

So admittedly, according to the spec, the "new" behaviour is correct. 
Should I submit a PR to remove the comment that the order is being preserved? I'm wondering if this should be considered a breaking change. 

Or do you have an opinion / better idea on what to do here? 

Thanks!

**Comment by tommyknows** [comment]
I'm just piling another small issue on top: #52 starts making use of `JoinPath`, which I've just noticed has only been added with Go 1.19. The `go.mod` file still only requires Go 1.17, so this just broke a build for me where a tool is using this library with Go 1.18...

I'd suggest to bump that version as well, or replace the usage of `JoinPath` with [it's implementation](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/net/url/url.go;l=1261):
```go
	url, err := Parse(base)
	if err != nil {
		return
	}
	result = url.JoinPath(elem...).String()
	return
```

Again, happy to open a PR for both of these issues!

**Comment by shibumi** [comment]
> Should I submit a PR to remove the comment that the order is being preserved? I'm wondering if this should be considered a breaking change.

Yes please. We should strictly follow the PURL spec and everything that differs, is clearly wrong.

**Comment by shibumi** [comment]
> I'm just piling another small issue on top: https://github.com/package-url/packageurl-go/pull/52 starts making use of JoinPath, which I've just noticed has only been added with Go 1.19. The go.mod file still only requires Go 1.17, so this just broke a build for me where a tool is using this library with Go 1.18...

We should only support the last 2 Go versions and maybe adapt our CICD that it tests the last two Go releases.

**Comment by shibumi** [comment]
PRs are welcome of course :)

**Comment by tommyknows** [comment]
Oh sorry, I hope your recovery is going well! 🍀 

Alright, I've opened #56 for this.

**Comment by petergardfjall** [comment]
After #83 the library should produce canonical purls with qualifiers ordered as per the spec. I'll close this. Feel free to repoen.
